### PR TITLE
open for inheritance

### PR DESCRIPTION
**Pull request by enpasos**
To implement an onnx export (see https://github.com/deepjavalibrary/djl/discussions/1420#discussioncomment-1931364) I would like to extend the export capability of existing block classes by inheritance. To do this, it would be very helpful if the DJL classes were opened up more to inheritance (the builders and the classes themselves).

This pull request is a suggestion on how this could happen.
Since it simply expands access with respect to inheritance, it should have no impact on existing implementations that use the library.

### CODE REVIEW

File: api/src/main/java/ai/djl/modality/cv/MultiBoxDetection.java [hunk 1]
```diff
@@ -37,12 +37,12 @@
  */
 public class MultiBoxDetection {
 
-    private boolean clip;
-    private float threshold;
-    private int backgroundId;
-    private float nmsThreshold;
-    private boolean forceSuppress;
-    private int nmsTopK;
+    protected boolean clip;
```
[the remark this answers is not in the document]
Comment (enpasos) [reply api/src/main/java/ai/djl/modality/cv/MultiBoxDetection.java]: My need is block inheritance: see https://github.com/enpasos/DJL2OnnxExample.
I reverted suggested changes that do not fall into this domain.

File: api/src/main/java/ai/djl/modality/nlp/DefaultVocabulary.java [hunk 2]
```diff
@@ -211,15 +211,15 @@ public static Builder builder() {
...
     /** Builder class that is used to build the {@link DefaultVocabulary}. */
-    public static final class Builder {
+    public static class Builder {
 
-        List<List<String>> sentences = new ArrayList<>();
-        Set<String> reservedTokens = new HashSet<>();
-        int minFrequency = -1;
-        int maxTokens = -1;
-        String unknownToken;
+        protected List<List<String>> sentences = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (enpasos) [reply api/src/main/java/ai/djl/modality/nlp/DefaultVocabulary.java]: My need is block inheritance: see https://github.com/enpasos/DJL2OnnxExample.
I reverted the suggested changes that do not fall into this domain.

**Comment by enpasos** [comment]
In the example I set up for onnx export I am facing the following specific inheritance problems:

grafik

From the screenshot you see a particular problem with Conv2d with package privacy on the constructors and the final Builder class.

You can see Conv2d, Linear, LayerNorm, LamdaBlock which I could not inherit easily. In the pull request I also included other classes with likely similar situations ... as I would not like to bother you daily with some more classes.

**Comment by enpasos** [comment]
To achieve an ONNX export for my MuZero implementation 

https://enpasos.ai/muzero/How#onnx

I only "opened" the four Conv2d, Linear, LayerNorm, LambdaBlock a little as mentioned above. 

What do you think about: 
- I close this Pull Request
- I open a new Pull Request with only changes in the four classes

That would help me in the way that I could remove the patches of the four classes from my working MuZero ONNX export - which likely is not the best way to do it in general, but at least works for this case.

**Comment by frankfliu** [comment]
@enpasos 

Sorry for the delay. I finally get some time to look into this. I look a bit in your repo, and I think the inherit solution doesn't provide a good user experience. It requires user to rewrite their model. I made some changes, it seems we can achieve the Onnx conversion by implementing a set of `OnnxConverter`, see: https://github.com/frankfliu/DJL2OnnxExample/tree/converter/onnxModelGen/src/main/java/ai/enpasos/mnist/blocks

Of course, we need make some changes to DJL Block API to expose some internal information. The change can be minimal: https://github.com/frankfliu/djl/commit/a1b789f2f7379b91ada7bc7e001591e4fe7db20a

Let me know what do you think.

**Comment by enpasos** [comment]
@frankfliu 

Thank you for your expert look into the subject. Your converter approach gives a nice separation of concerns - much better than the inheritance approach I used. Looks promissing to achieve a good UX.

My approach was to see if I could manage to run MuZero Inference in the browser. Result: Works fine on the CPU via WebAssembly, but not on the GPU via WebGL, because there were too few network functions supported in WebGL.

I would like to add some thoughts from the experience I made on that way:

- Translation of the network blocks was not always straight forward. For example the translation of the layer norm was not nice because there is not yet a corresponding onnx component. I ended up with a far from nice solution that gave the right mathematical behaviour misusing a batchnorm component. From this experience I deduce that it is important to **allow for future replacement of particular converters**.
- It could also be that the target system where the onnx should be used can not work with a particular onnx component or version. That also suggests to **give the end user the freedom to exchange a converter implementation if needed**.
-> I think this could be achieved by **allowing the user to provide an implementation of the factory method** `OnnxConverter getConverter(Block block)`

- the ParallelBlock takes an input for the join function. The solutions needs the flexibility that the writer of the join function can also provide the corresponding translation to onnx.

- I got a [comment by Craigacp](https://github.com/enpasos/DJL2OnnxExample/issues/1) who pointed out that they worked on an onnx solution in Tribuo that might help here.  I didn't switch because I almost achieved my MuZero export goal at that time.

**Comment by enpasos** [comment]
@frankfliu

For me it would be perfect to switch to your approach. 

To be able to map an onnx join function to a djl join function in ParallelBlocks it would be nice to have the option to provide a name string for the parallel function that can be used for the mapping.

I do not know what would be the best way to proceed. If you like you can close and open a new issue, or rename this issue. 

What do you think?

**Comment by enpasos** [comment]
Two reasons why I closed the old issue and opened a new one (https://github.com/deepjavalibrary/djl/pull/2231):
-  I was reorganizing the branches of my fork
-  I reduced the content of the change to what I really need

**Pull request closed without merging**